rfd/0131-adminitrative-actions-mfa.md

```diff
@@ -0,0 +1,260 @@
+- Adding, removing, or editing users, roles, auth connectors, and other
+administrative resources.
+- Approving access requests.
+
+This will apply to administrative actions performed via the WebUI, `tctl`,
+and `tsh`.
+
+## Why
+
+Currently, Teleport only prompts for MFA verification in a few circumstances:
+
+1. User logs in with MFA.
+2. User updates their password or registered MFA devices.
+3. User connects to a service that requires Per-session MFA.
+
+As a result, the vast majority of Teleport actions do not require timely MFA
+verification. If an admin's login session is compromised, it can be used to
+perform attacks pertaining to privilege escalation, security downgrades, Auth
+connector changes, and other forms of cluster sabotage.
+
+Adding an MFA restriction to administrative actions limits this attack vector
+by re-verifying the user's identity promptly before performing any administrative
+action that could be abused.
+
+## Details
+
+### Server configuration
+
+Starting in Teleport 14, MFA for administrative actions will be turned on by
+default for all users with an MFA device registered. This means that when
+a cluster's second factor is set to one of `optional`, `on`, `webauthn`, or
+`otp`, all normal login sessions will require MFA verification to complete
+admin actions.
+
+### Administrative Actions
+
+An administrative action will be primarily defined as an action which modifies a
+resource in a Teleport Cluster directly. Essentially, this covers any action which
+requires an allow rule with a `create`, `update`, or `delete` verb requirement.
+
+For example:
+
+- `tctl rm/create/edit`
+- `tctl users add/rm/reset`
+- Adding a node from the WebUI
+- Modifying a role form the WebUI
+
+#### Usage Event Creation
+
+The `usage_event` resource can be created by any Teleport User via the
+`DefaultImplicitRole`. This will not be considered an admin action.
+
+### Access Request Approval
+
+Submitting an access request approval does not require an `update` allow rule,
+but since it can be used to escalate privileges, it will still be considered an
+admin action.
+
+Submitting an access request denial will remain a non-admin action.
+
+#### User Actions
+
+Some actions which create, update, or delete user specific resources are
+authorized based on the identity of the requesting user rather than their
+allow rules. For example:
+
+- Creating, listing, and deleting a user's own access requests.
+- Adding or removing MFA devices.
+- Changing the user's password.
+
+These actions will continue to be non-administrative actions and will not require
+MFA to complete.
+
+### UX
+
+The UX of this feature will be very similar to Per-session MFA for `tsh`, `tctl`,
+the WebUI, and Teleport Connect.
+
+When a user performs an admin action, they will be prompted to tap their MFA key.
+Each admin action should require a single tap.
+
+```console
+$ tctl edit roles/access
+# edit role in vim
+Tap any security key
+# success
```

Comment:
This flow is probably easiest to implement, but the UX kind of stinks if the MFA fails for any reason.

For example, if you make a bunch of changes to the role and mistakenly tap the wrong MFA device, your changes would be lost and you'd have to repeat the whole process.

Thoughts for how we might improve this? Do we care? Maybe we treat `tctl edit` as a special case and prompt for MFA _first_?

Reply:
I agree that prompting for MFA first is best, and it's been a consideration for me, but I left it like this for simplicity as this is how it will be for most commands.

However, we can probably just carve out `tctl edit` as an edge case and just always prompt for MFA up front. I will give this more thought in the next draft of the RFD.